OCPBUGS-33285: Add TelemeterClientFailures runbook (#203)
openshift/runbooks: alerts/cluster-monitoring-operator/TelemeterClientFailures.md

# TelemeterClientFailures

## Meaning

The alert `TelemeterClientFailures` is triggered when the Telemeter client

> fails to send metrics to the Telemeter server.

Reviewer: ```suggestion
fails to send metrics to Red Hat.

The `telemeter-client-xxxxxxxxx-xxxxx` pod running in the `openshift-monitoring`

> namespace collects [selected platform metrics](https://docs.openshift.com/container-platform/latest/support/remote_health_monitoring/showing-data-collected-by-remote-health-monitoring.html#showing-data-collected-from-the-cluster_showing-data-collected-by-remote-health-monitoring)
> from the `prometheus-k8s` service at
> regular intervals(`4m30s`) using the `/federate` endpoint and ships the samples
> to the Telemetry endpoint using a custom protocol.

Reviewer: (nit) I wouldn't give concrete numbers to avoid discrepancy in the future. 
```suggestion
namespace collects [selected platform metrics](https://docs.openshift.com/container-platform/latest/support/remote_health_monitoring/showing-data-collected-by-remote-health-monitoring.html#showing-data-collected-from-the-cluster_showing-data-collected-by-remote-health-monitoring)
from the `prometheus-k8s` service  and ships them periodically
to Red Hat using a custom secured protocol.
```

## Impact

> When `telemeter-client` is having failures it will not be able to send metrics
> to the [telemetry system](https://docs.openshift.com/container-platform/latest/support/remote_health_monitoring/about-remote-health-monitoring.html#telemetry-about-telemetry_about-remote-health-monitoring).

Reviewer: (suggestion)
```suggestion
When the alert fires, Red Hat support and engineering teams don't have a complete view of the cluster anymore. It may hinder the ability for Red Hat to proactively detect issues in the cluster.
```

> ## Diagnosis

Reviewer: We should probably break into 2 options:
* Telemeter client fails to pull metrics from Prometheus.
  * You'd need to check the availability of Prometheus.
* Telemeter client fails to push to the server.
  * It can be a server outage
  * It can be an issue with the cluster pull secret.

Review the logs for the pod `telemeter-client-xxxxxxxxx-xxxxx`
in the `openshift-monitoring` namespace

You can review the logs for the `telemeter-client-xxxxxxxxx-xxxxx` pod in the
`openshift-monitoring` namespace by running the following command:

Replace `telemeter-client-xxxxxxxxx-xxxxx` with exact telemeter-client pod name

```console

> $ oc logs telemeter-client-xxxxxxxxx-xxxxx -c telemeter-client | less

Reviewer: Can we do `oc logs  deployment/telemeter-client ...`

> $ oc logs telemeter-client-xxxxxxxxx-xxxxx -c telemeter-client | less

Reviewer 2: `oc logs -n openshift-monitoring deployment.apps/telemeter-client -c telemeter-client`

```

## Mitigation

* If the telemeter client fails to authenticate with the telemeter service,
make sure that the global pull secret is up to date, see

> [here](https://docs.openshift.com/container-platform/latest/openshift_images/managing_images/using-image-pull-secretshtml#images-update-global-pull-secret_using-image-pull-secrets)

Reviewer 2: ```suggestion
[here](https://docs.openshift.com/container-platform/latest/openshift_images/managing_images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets)
```

# TelemeterClientFailures

## Meaning

> The alert `TelemeterClientFailures` is triggered when the Telemeter client tries
> and fails to send Telemetry data at a certain rate over a period of time
> to Red Hat.

Reviewer: ```suggestion
The alert `TelemeterClientFailures` is triggered when the Telemeter client
fails to send Telemetry data at a certain rate over a period of time
to Red Hat.
```

from the `prometheus-k8s` service at
regular intervals using the `/federate` endpoint and ships them periodically
to Red Hat using a custom secured protocol.

## Impact

When the alert fires, Red Hat support and engineering teams don't have a complete
view of the cluster anymore. It may hinder the ability for Red Hat to
proactively detect issues in the cluster.

## Diagnosis

Review the logs for the pod `telemeter-client-xxxxxxxxx-xxxxx`
in the `openshift-monitoring` namespace

You can review the logs for the `telemeter-client-xxxxxxxxx-xxxxx` pod in the
`openshift-monitoring` namespace by running the following command:

```console
$ oc logs -n openshift-monitoring deployment.apps/telemeter-client -c telemeter-client | less
```

## Mitigation

The resolution depends on the particular issue reported in the logs.

* Telemeter client fails to pull metrics from Prometheus.
  * You'd need to check the availability of Prometheus.

* Telemeter client fails to push to the server.

  * It can be due to Red Hat Telemeter server outage.

    ```console
    2024-05-05T09:33:28.347041825+08:00 level=error caller=forwarder.go:276 ts=2024-05-05T01:33:28.346019008Z

>     component=forwarder/worker msg="unable to forward results" err="gateway server reported unexpected error code: 503:

Reviewer: I'd emphasize that u`nexpected error code` means an error on the server side (especially with a 5xx status code).

Otherwise status codes like 401 or 403 are signs of an invalid pull secret.

# TelemeterClientFailures

## Meaning

The alert `TelemeterClientFailures` is triggered when the Telemeter client fails
to send Telemetry data at a certain rate over a period of time
to Red Hat.

> The `telemeter-client-xxxxxxxxx-xxxxx` pod running in the `openshift-monitoring`

Reviewer 2: ```suggestion
The `telemeter-client` pod running in the `openshift-monitoring`
```
I think `telemeter-client` is enough, same for other similar sentences

Author: But exact pod name isn't there thats why added the hash

Reviewer: I think that "Telemeter client pod" would be ok.

namespace collects [selected platform metrics](https://docs.openshift.com/container-platform/latest/support/remote_health_monitoring/showing-data-collected-by-remote-health-monitoring.html#showing-data-collected-from-the-cluster_showing-data-collected-by-remote-health-monitoring)
from the `prometheus-k8s` service at
regular intervals using the `/federate` endpoint and ships them periodically
to Red Hat using a custom secured protocol.

## Impact

When the alert fires, Red Hat support and engineering teams don't have a complete
view of the cluster anymore. It may hinder the ability for Red Hat to
proactively detect issues in the cluster.

## Diagnosis

Review the logs for the pod `telemeter-client-xxxxxxxxx-xxxxx`

> in the `openshift-monitoring` namespace

Reviewer: ```suggestion
in the `openshift-monitoring` namespace.
```

You can review the logs for the `telemeter-client-xxxxxxxxx-xxxxx` pod in the
`openshift-monitoring` namespace by running the following command:

```console
$ oc logs -n openshift-monitoring deployment.apps/telemeter-client -c telemeter-client | less
```

## Mitigation

> * From `4.16` there is metric `metricsclient_http_requests_total` which can be

Reviewer 2: `metricsclient_http_requests_total`is since 4.17
```
$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.17.0-0.nightly-2024-08-13-031847   True        False         7h23m   Cluster version is 4.17.0-0.nightly-2024-08-13-031847

$ token=`oc create token prometheus-k8s -n openshift-monitoring`
$ oc -n openshift-monitoring exec -c prometheus prometheus-k8s-0 -- curl -k -H "Authorization: Bearer $token" 'https://thanos-querier.openshift-monitoring.svc:9091/api/v1/label/__name__/values' | jq | grep metricsclient
    "metricsclient_http_requests_total",
    "metricsclient_request_retrieve",
    "metricsclient_request_send",
```
not in 4.16
```
$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.16.0-0.nightly-2024-08-14-124952   True        False         5h46m   Cluster version is 4.16.0-0.nightly-2024-08-14-124952
$ token=`oc create token prometheus-k8s -n openshift-monitoring`
$ oc -n openshift-monitoring exec -c prometheus prometheus-k8s-0 -- curl -k -H "Authorization: Bearer $token" 'https://thanos-querier.openshift-monitoring.svc:9091/api/v1/label/__name__/values' | jq | grep metricsclient
    "metricsclient_request_retrieve",
    "metricsclient_request_send",
```

Author: right my bad :)

  used to identify which side the issue is happening. For older cluster versions

>  `metricsclient_request_retrieve` and `metricsclient_request_send` metrices

Reviewer: ```suggestion
 `metricsclient_request_retrieve` and `metricsclient_request_send` metrics
```

>   can be used to check the server status.

Reviewer: I would give PromQL expressions to guide the users (not tested):
* 4.17+
  * `sum by(client, status_code) (rate(metricsclient_http_requests_total{status_code!~"2.."})) > 0` 
  * Explain that the "client" label value shows which side is faulty
    * `federate_from` when the error is from Prometheus
    * `federate_to` when the error is from Telemeter server
* 4.16 and below
  * `sum by(client, status_code) (rate(metricsclient_request_retrieve{status_code!~"2.."})) > 0` (Prometheus)
  * `sum by(client, status_code) (rate(metricsclient_request_send{status_code!~"2.."})) > 0` (Telemeter server)

Also we're still in `Diagnosis` here.

Author: Checked the query in console its giving error, working on it

Author: Issue was 

```
1:35: parse error: expected type range vector in call to function "rate", got instant vector
```

updated query to range vector also `!~"2.."` was not working I tried with `=` as well just to make sure some results return

The resolution depends on the particular issue reported in the logs.

* Telemeter client fails to pull metrics from Prometheus.
  * You'd need to check the availability of Prometheus.

* Telemeter client fails to push to the server.

>   * It can be due to Red Hat Telemeter server outage.

Reviewer 2: ```suggestion
  * It can be due to the Red Hat Telemeter server outage.
```

in the `openshift-monitoring` namespace.

You can review the logs for the `telemeter-client` pod in the
`openshift-monitoring` namespace by running the following command:

```console
$ oc logs -n openshift-monitoring deployment.apps/telemeter-client -c telemeter-client | less
```

* Execute the PromQL expressions to identify which side issue is happening

  * 4.17+

    ```console
    sum by(client, status_code) (rate(metricsclient_http_requests_total{status_code!~"2.."})) > 0
    ```

    * `client` label value shows which side is faulty.
    * `federate_from` when the error is from Prometheus.
    * `federate_to` when the error is from Telemeter server.

  * 4.16 and below

    Check Prometheus status

      ```console
      sum by(client, status_code) (rate(metricsclient_request_retrieve{status_code!~"2.."})) > 0
      ```

    Check Telemeter Server status

      ```console
      sum by(client, status_code) (rate(metricsclient_request_send{status_code!~"2.."})) > 0
      ```

## Mitigation

The resolution depends on the particular issue reported in the logs.

* Telemeter client fails to pull metrics from Prometheus.
  * You'd need to check the availability of Prometheus.

> * Telemeter client fails to push to the server.

Reviewer: worth mentioning that HTTP proxies need to allow with the telemeter server names (https://docs.openshift.com/container-platform/4.15/installing/install_config/configuring-firewall.html).

```

* Execute the PromQL expressions to identify which side issue is happening

  * 4.17+

    ```console
    sum by(client, status_code) (rate(metricsclient_http_requests_total{status_code!~"2.."})) > 0
    ```

    * `client` label value shows which side is faulty.
    * `federate_from` when the error is from Prometheus.
    * `federate_to` when the error is from Telemeter server.

  * 4.16 and below

    Check Prometheus status

      ```console
      sum by(client, status_code) (rate(metricsclient_request_retrieve{status_code!~"200"}[5m])) > 0
      ```

    Check Telemeter Server status

      ```console
      sum by(client, status_code) (rate(metricsclient_request_send{status_code!~"200"}[5m])) > 0
      ```

## Mitigation

The resolution depends on the particular issue reported in the logs.

* Telemeter client fails to pull metrics from Prometheus.
  * You'd need to check the availability of Prometheus.

* Telemeter client fails to push to the server.

  * Make sure [HTTP proxies allow](https://docs.openshift.com/container-platform/latest/installing/install_config/configuring-firewall.html)
  the telemeter server names.

>   * It can be due to Red Hat Telemeter server outage.

Reviewer: Should we ask users to check the `status_code` label value as it would be more convenient that looking through the logs?
Another follow-up improvement could be to have new alerting rules based on the `metricsclient_http_requests_total` metric which would tell the difference between 4xx errors (e.g. bad pull secret) and 5xx (issue on Red Hat side).

Author: ya I can create a separate jira for that as this one is for runbook

Reviewer: https://github.com/prometheus-operator/prometheus-operator/issues/6948
